### ISSUE TITLE
how to manipulate the arrived messages ?

### ISSUE THREAD

**Issue by vinidg** [walterchau/ionic-mqtt#3]
How do I assign the value of messages to a variable to manipulate my ionic objects?

**Comment by vinidg** [comment]
Any can help me ?

**Comment by 0x-KChau** [comment]
> Any can help me ?

It would be great if you can provide your source code for better understanding.

**Comment by vinidg** [comment]
I'm sorry I didn't explain very well. In incoming messages method I can not update device status, the result is the error below

```
  private _onMessageArrived(message) {
    console.log("message", message);
    let topico = message.topic;
    let payload = message.payloadString;
    let macAddress = topico.split('/', 3)[0];
    let statusRede = topico.split('/', 3)[2];

    this.dispositivos = this.dispositivos.map(device => {
      if(device.Dispositivo.MacAddress === macAddress) {
          if(statusRede === this.STATUS_REDE){
            device.Dispositivo.StatusRede = payload;
          }else if(statusRede === this.STATUS) {
            device.Dispositivo["turn"] = payload;
        }
      }
      })
    }
  }
```
The error: 

```
_onConnectionLost:
errorCode 5
"AMQJS0005E Internal error. Error Message: Cannot read property 'map' of undefined, Stack trace: No Error Stack Available"
```

**Comment by 0x-KChau** [comment]
> "AMQJS0005E Internal error. Error Message: Cannot read property 'map' of undefined, Stack trace: No Error Stack Available"

It looks like it breaks because of "this.dispositivos" is undefined. Please correct me if I am wrong.

**Comment by vinidg** [comment]
Yes, you're right. But this variable is not undefined. This happens when I use it within the onMessageArrived method.

**Comment by vinidg** [comment]
I need an example of how you did to implement this part

**Comment by 0x-KChau** [comment]
> I need an example of how you did to implement this part

The basic implementation is just like this.
```
private _onMessageArrived(message) {
    console.log("message", message);
    //...do with the message
  }
```
Small reminder, when using Array.prototype.map(), you have to return sth, otherwise it will breaks.

**Comment by vinidg** [comment]
ok, but if I log my page variables, and the function is executed it returns them all as undefined.  This example implementation works fine, but I need to use the returned message somehow

**Comment by 0x-KChau** [comment]
> ok, but if I log my page variables, and the function is executed it returns them all as undefined. This example implementation works fine, but I need to use the returned message somehow

What I mean is you have to return sth inside your map function.
```
    this.dispositivos = this.dispositivos.map(device => {
      if(device.Dispositivo.MacAddress === macAddress) {
          if(statusRede === this.STATUS_REDE){
            return device.Dispositivo.StatusRede = payload;
          }else if(statusRede === this.STATUS) {
             return device.Dispositivo["turn"] = payload;
        }
      }
      })
```

**Comment by vinidg** [comment]
I understand, but my problem is another

![image](https://user-images.githubusercontent.com/10141812/72207069-c6d5b880-3473-11ea-9d1f-559af4eb843d.png)

**Comment by robingrandval** [comment]
Hi Vinidg,

Is due to your context. "this" on your callback function ("handleStatusDevice") doesn't represent your angular component "dispositivos.page", so this.dispositivos doesn't exist.

Hope isn't too late :)
